File: spring-boot-domain-persistence-modeling-java/src/main/java/x/scratch/UpsertableRecord.java
package x.scratch;

import lombok.RequiredArgsConstructor;
import lombok.Value;

import java.util.Optional;

import static lombok.AccessLevel.PRIVATE;

public interface UpsertableRecord<Record extends UpsertableRecord<Record>> {
    Record upsertedWith(Record upserted);

    @RequiredArgsConstructor(access = PRIVATE)
    @Value
    final class UpsertedRecordResult<Record extends UpsertableRecord<Record>> {
        private final Record record;
        private final boolean changed;

        public static <Record extends UpsertableRecord<Record>> UpsertedRecordResult<Record> of(
                final Record entity, final Optional<Record> upserted) {
            return new UpsertedRecordResult<>(entity, upserted.isPresent());
        }
    }
}

File: domain-persistence-modeling/kotlin-spring-boot/src/main/kotlin/x/domainpersistencemodeling/UpsertableRecord.kt
package x.domainpersistencemodeling

import java.util.Optional

interface UpsertableRecord<Record : UpsertableRecord<Record>> {
    fun upsertedWith(upserted: Record): Record

    data class UpsertedRecordResult<Record : UpsertableRecord<Record>>(
            val record: Record, val changed: Boolean) {
        constructor(record: Record, upserted: Optional<Record>)
                : this(record, upserted.isPresent)
    }
}
